include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -152,13 +152,27 @@ __pattern_transform_scan_base_impl(_ExecutionPolicy&& __exec, _Iterator1 __first
     }
 }
 
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct is_equality_implemented : std::false_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct is_equality_implemented<_Iterator1, _Iterator2,
+                               std::void_t<decltype(std::declval<::std::decay_t<_Iterator1>>().operator==(
+                                   std::declval<::std::decay_t<_Iterator2>>()))>> : std::true_type
+{
+};
+
 template <typename _Iterator1, typename _Iterator2>
 constexpr bool
 __check_equal_iterators(_Iterator1 __it1, _Iterator2 __it2)
 {
-    // In-place exclusive scan works correctly only if an input and an output iterators are the same type.
+    // In-place exclusive scan works correctly only if an input and an output iterators are the same type
+    // of operator== exist for their type - so they are comparable
     // Otherwise, there is no way to check an in-place case and a workaround below is not applied.
-    if constexpr (::std::is_same_v<::std::decay_t<_Iterator1>, ::std::decay_t<_Iterator2>>)
+    if constexpr (::std::is_same_v<::std::decay_t<_Iterator1>, ::std::decay_t<_Iterator2>> ||
```

Comment:
whether is not enough just `is_equality_implemented` here?

Reply:
Fixed: now ```is_equality_implemented``` implemented and used only in one file